packages/connect-core/src/serialization.ts

```diff
@@ -0,0 +1,138 @@
+  BinaryWriteOptions,
+  JsonReadOptions,
+  JsonWriteOptions,
+  Message,
+  MessageType,
+  MethodInfo,
+  PartialMessage,
+} from "@bufbuild/protobuf";
+import { ConnectError } from "./connect-error.js";
+import { Code } from "./code.js";
+
+/**
+ * Serialization provides methods to serialize or parse data with a certain
+ * format.
+ */
+export interface Serialization<T> {
+  serialize: (data: T) => Uint8Array;
+
+  parse: (data: Uint8Array) => T;
+}
+
+/**
+ * Options for createBinarySerialization()
+ */
+export type BinarySerializationOptions = Partial<
+  BinaryReadOptions & BinaryWriteOptions
+>;
+
+/**
+ * Creates a Serialization object for serializing the given protobuf message
+ * with the protobuf binary format.
+ */
+export function createBinarySerialization<T extends Message<T>>(
+  messageType: MessageType<T>,
+  options: BinarySerializationOptions | undefined
+): Serialization<T> {
+  return {
+    parse(data: Uint8Array): T {
+      try {
+        return messageType.fromBinary(data, options);
+      } catch (e) {
+        const m = e instanceof Error ? e.message : String(e);
+        throw new ConnectError(`parse binary: ${m}`, Code.InvalidArgument);
+      }
+    },
+    serialize(data: T): Uint8Array {
+      try {
+        return data.toBinary(options);
+      } catch (e) {
+        const m = e instanceof Error ? e.message : String(e);
+        throw new ConnectError(`serialize binary: ${m}`, Code.Internal);
+      }
+    },
+  };
+}
+
+/**
+ * Options for createJsonSerialization()
+ */
+export type JsonSerializationOptions = Partial<
+  JsonReadOptions & JsonWriteOptions
+> & {
+  textEncoder?: { encode(input?: string): Uint8Array };
+  textDecoder?: { decode(input?: Uint8Array): string };
+};
+
+/**
+ * Creates a Serialization object for serializing the given protobuf message
+ * with the protobuf canonical JSON encoding.
+ */
+export function createJsonSerialization<T extends Message<T>>(
+  messageType: MessageType<T>,
+  options: JsonSerializationOptions | undefined
+): Serialization<T> {
+  const textEncoder = options?.textEncoder ?? new TextEncoder();
+  const textDecoder = options?.textDecoder ?? new TextDecoder();
+  return {
+    parse(data: Uint8Array): T {
+      try {
+        const json = textDecoder.decode(data);
+        return messageType.fromJsonString(json, options);
+      } catch (e) {
+        const m = e instanceof Error ? e.message : String(e);
+        throw new ConnectError(m, Code.InvalidArgument);
+      }
+    },
+    serialize(data: T): Uint8Array {
+      try {
+        const json = data.toJsonString(options);
+        return textEncoder.encode(json);
+      } catch (e) {
+        const m = e instanceof Error ? e.message : String(e);
+        throw new ConnectError(m, Code.Internal);
+      }
+    },
+  };
+}
+
+/**
+ * Returns functions to normalize and serialize the input message
+ * of an RPC, and to parse the output message of an RPC.
+ */
+export function createClientMethodSerializers<
```

Comment:
It still makes sense to keep this convenience-function. We need to update [createServerMethodsSerializers](https://github.com/bufbuild/connect-web/blob/1bc26cae5c95e3f41fd63dfa90707f207176b595/packages/connect-node/src/private/create-server-method-serializers.ts) as well.